File: capy/src/test/js/capy.test.js
const test = require('node:test');
const assert = require('node:assert/strict');
const { mkdtemp, mkdir, writeFile, readFile, access } = require('node:fs/promises');
const { spawnSync } = require('node:child_process');
const { join } = require('node:path');
const { tmpdir } = require('node:os');

function runCapy(args) {
    const classpath = process.env.CAPY_CLI_CLASSPATH;
    assert.ok(classpath, 'CAPY_CLI_CLASSPATH must be provided by the Gradle jsTests task');
    return spawnSync('java', ['-cp', classpath, 'dev.capylang.Capy', ...args], {
        encoding: 'utf8',
    });
}

function runNode(args) {
    return spawnSync('node', args, {
        encoding: 'utf8',
    });
}

async function tempProject() {
    return mkdtemp(join(tmpdir(), 'capy-js-test-'));
}

async function exists(path) {
    try {
        await access(path);
        return true;
    } catch {
        return false;
    }
}

test('prints version', () => {
    const result = runCapy(['--version']);

    assert.equal(result.status, 0, result.stderr);
    assert.equal(result.stderr.trim(), '');
    assert.match(result.stdout.trim(), /^Capybara compiler version: /);
});

test('compile-generate JS writes linked JSON, CommonJS modules, and runnable output', async () => {
    const root = await tempProject();
    const sourceDir = join(root, 'src');
    const generatedDir = join(root, 'generated');
    const linkedDir = join(root, 'linked');
    await mkdir(join(sourceDir, 'foo'), { recursive: true });
    await writeFile(join(sourceDir, 'foo', 'Main.cfun'), `
from /capy/lang/Option import { * }
from /capy/collection/List import { * }

fun answer(): int = [10, 20, 12].reduce(0, (acc, value) => acc + value)

fun maybe(flag: bool): Option[String] =
    if flag then Some { "ok" } else None {}

fun read(flag: bool): String =
    match maybe(flag) with
    case Some { value } -> value
    case None -> "none"
`);

    const result = runCapy([
        'compile-generate',
        'JS',
        '-i', sourceDir,
        '-o', generatedDir,
        '--linked-output', linkedDir,
    ]);

    assert.equal(result.status, 0, result.stderr);
    assert.equal(result.stdout.trim(), '');
    assert.equal(await exists(join(linkedDir, 'program.json')), true);
    assert.equal(await exists(join(linkedDir, 'foo', 'Main.json')), true);
    assert.equal(await exists(join(generatedDir, 'foo', 'Main.js')), true);
    assert.equal(await exists(join(generatedDir, 'dev', 'capylang', 'capybara.js')), true);

    const node = runNode([
        '-e',
        `const m = require(${JSON.stringify(join(generatedDir, 'foo', 'Main.js'))}); console.log([m.answer(), m.read(true), m.read(false)].join('|'));`,
    ]);
    assert.equal(node.status, 0, node.stderr);
    assert.equal(node.stdout.trim(), '42|ok|none');
});

test('generate JS from linked input', async () => {
    const root = await tempProject();
    const sourceDir = join(root, 'src');
    const linkedDir = join(root, 'linked');
    const generatedDir = join(root, 'generated');
    await mkdir(join(sourceDir, 'foo'), { recursive: true });
    await writeFile(join(sourceDir, 'foo', 'Main.cfun'), 'fun answer(): int = 7\n');

    const compile = runCapy(['compile', '-i', sourceDir, '-o', linkedDir]);
    assert.equal(compile.status, 0, compile.stderr);

    const generate = runCapy(['generate', 'javascript', '-i', linkedDir, '-o', generatedDir]);
    assert.equal(generate.status, 0, generate.stderr);
    assert.equal(await exists(join(generatedDir, 'foo', 'Main.js')), true);

    const node = runNode([
        '-e',
        `const m = require(${JSON.stringify(join(generatedDir, 'foo', 'Main.js'))}); console.log(m.answer());`,
    ]);
    assert.equal(node.status, 0, node.stderr);
    assert.equal(node.stdout.trim(), '7');
});

test('compile-generate JS accepts native provider annotation wiring', async () => {
    const root = await tempProject();
    const sourceDir = join(root, 'src');
    const generatedDir = join(root, 'generated');
    const linkedDir = join(root, 'linked');
    const regeneratedDir = join(root, 'regenerated');
    await mkdir(join(sourceDir, 'foo'), { recursive: true });
    await mkdir(join(sourceDir, 'dev', 'capylang', 'test'), { recursive: true });
    await writeFile(join(sourceDir, 'foo', 'Main.cfun'), 'fun answer(): int = 9\n');
    await writeFile(join(sourceDir, 'dev', 'capylang', 'test', 'Clock.coo'), `
interface Clock {
    def now(): String
}
`);
    await writeFile(join(sourceDir, 'dev', 'capylang', 'test', 'ClockProvider.cfun'), `
from /capy/lang/Effect import { Effect }
from /capy/meta_prog/NativeProvider import { NativeProvider }
from Clock import { Clock }

@NativeProvider(
    qualifier: "system",
    lifetime: "factory",
    javascriptModule: "./nativeinterop/system_clock.js",
    javascriptExport: "SystemClock"
)
fun system_clock(): Effect[Clock] = <native>
`);

    const result = runCapy([
        'compile-generate',
        'JS',
        '-i', sourceDir,
        '-o', generatedDir,
        '--linked-output', linkedDir,
    ]);

    assert.equal(result.status, 0, result.stderr);
    const programJson = await readFile(join(linkedDir, 'program.json'), 'utf8');
    assert.match(programJson, /nativeProviders/);
    assert.match(programJson, /nativeProviderCatalog/);
    assert.match(programJson, /nativeinterop\/system_clock\.js/);
    assert.equal(await exists(join(generatedDir, 'dev', 'capylang', 'native_providers.js')), true);

    const generate = runCapy(['generate', 'js', '-i', linkedDir, '-o', regeneratedDir]);
    assert.equal(generate.status, 0, generate.stderr);
    assert.equal(await exists(join(regeneratedDir, 'foo', 'Main.js')), true);
    assert.equal(await exists(join(regeneratedDir, 'dev', 'capylang', 'native_providers.js')), true);
});

test('compile-generate JS rejects malformed native wiring manifest', async () => {
    const root = await tempProject();
    const sourceDir = join(root, 'src');
    const generatedDir = join(root, 'generated');
    const nativeWiringFile = join(root, 'capy.native.json');
    await mkdir(sourceDir, { recursive: true });
    await writeFile(join(sourceDir, 'Main.cfun'), 'fun answer(): int = 1\n');
    await writeFile(nativeWiringFile, '{ not json');

    const result = runCapy([
        'compile-generate',
        'js',
        '-i', sourceDir,
        '-o', generatedDir,
        '--native-wiring', nativeWiringFile,
    ]);

    assert.equal(result.status, 1);
    assert.equal(result.stdout.trim(), '');
    assert.match(result.stderr, /Unable to read native wiring manifest/);
    assert.match(result.stderr, new RegExp(nativeWiringFile.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')));
});

test('compile-generate JS accepts object-oriented modules', async () => {
    const root = await tempProject();
    const sourceDir = join(root, 'src');
    const generatedDir = join(root, 'generated');
    await mkdir(join(sourceDir, 'foo'), { recursive: true });
    await writeFile(join(sourceDir, 'foo', 'Main.coo'), `
class Main {
    def main(args: List[String]): int = args.size()
}
`);

    const result = runCapy(['compile-generate', 'JS', '-i', sourceDir, '-o', generatedDir]);

    assert.equal(result.status, 0, result.stderr);
    assert.equal(await exists(join(generatedDir, 'foo', 'Main.js')), true);

    const node = runNode([join(generatedDir, 'foo', 'Main.js'), 'one', 'two']);
    assert.equal(node.status, 0, node.stderr);
    assert.equal(node.stdout.trim(), '');
    assert.equal(node.stderr.trim(), '');
});

test('generated JS output is pruned by manifest on reuse', async () => {
    const root = await tempProject();
    const sourceDir = join(root, 'src');
    const generatedDir = join(root, 'generated');
    await mkdir(join(sourceDir, 'foo'), { recursive: true });
    await mkdir(join(generatedDir, 'stale'), { recursive: true });
    await writeFile(join(sourceDir, 'foo', 'Main.cfun'), 'fun answer(): int = 1\n');
    await writeFile(join(generatedDir, 'stale', 'Old.js'), 'module.exports = {};\n');

    const result = runCapy(['compile-generate', 'JS', '-i', sourceDir, '-o', generatedDir]);

    assert.equal(result.status, 0, result.stderr);
    assert.equal(await exists(join(generatedDir, 'foo', 'Main.js')), true);
    assert.equal(await exists(join(generatedDir, 'stale', 'Old.js')), false);
    const manifest = await readFile(join(generatedDir, '.capy-output-manifest'), 'utf8');
    assert.equal(manifest.includes('foo/Main.js'), true);
});

File: capy/src/test/py/capy_test.py
import os
import pathlib
import shutil
import subprocess
import sys
import tempfile
import textwrap
import unittest


def run_capy(args):
    classpath = os.environ["CAPY_CLI_CLASSPATH"]
    return subprocess.run(
        ["java", "-cp", classpath, "dev.capylang.Capy", *args],
        text=True,
        capture_output=True,
        check=False,
    )


def run_python(args, cwd=None):
    return subprocess.run(
        [sys.executable, *args],
        cwd=cwd,
        text=True,
        capture_output=True,
        check=False,
    )


class CapyPythonCliTest(unittest.TestCase):
    def temp_project(self):
        return pathlib.Path(tempfile.mkdtemp(prefix="capy-py-test-"))

    def test_compile_generate_python_writes_linked_json_and_runnable_output(self):
        root = self.temp_project()
        source_dir = root / "src"
        generated_dir = root / "generated"
        linked_dir = root / "linked"
        (source_dir / "foo").mkdir(parents=True)
        (source_dir / "foo" / "Main.cfun").write_text(textwrap.dedent("""
            from /capy/lang/Option import { * }
            from /capy/collection/List import { * }

            fun answer(): int = [10, 20, 12].reduce(0, (acc, value) => acc + value)

            fun maybe(flag: bool): Option[String] =
                if flag then Some { "ok" } else None {}

            fun read(flag: bool): String =
                match maybe(flag) with
                case Some { value } -> value
                case None -> "none"
        """))

        result = run_capy([
            "compile-generate",
            "PY",
            "-i", str(source_dir),
            "-o", str(generated_dir),
            "--linked-output", str(linked_dir),
        ])

        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(result.stdout.strip(), "")
        self.assertTrue((linked_dir / "program.json").exists())
        self.assertTrue((linked_dir / "foo" / "Main.json").exists())
        self.assertTrue((generated_dir / "foo" / "Main.py").exists())
        self.assertTrue((generated_dir / "dev" / "capylang" / "capybara.py").exists())

        python = run_python([
            "-c",
            "import foo.Main as m; print('|'.join([str(m.answer()), m.read(True), m.read(False)]))",
        ], cwd=generated_dir)
        self.assertEqual(python.returncode, 0, python.stderr)
        self.assertEqual(python.stdout.strip(), "42|ok|none")

    def test_generate_python_from_linked_input(self):
        root = self.temp_project()
        source_dir = root / "src"
        linked_dir = root / "linked"
        generated_dir = root / "generated"
        (source_dir / "foo").mkdir(parents=True)
        (source_dir / "foo" / "Main.cfun").write_text("fun answer(): int = 7\n")

        compile_result = run_capy(["compile", "-i", str(source_dir), "-o", str(linked_dir)])
        self.assertEqual(compile_result.returncode, 0, compile_result.stderr)

        generate = run_capy(["generate", "python", "-i", str(linked_dir), "-o", str(generated_dir)])
        self.assertEqual(generate.returncode, 0, generate.stderr)
        self.assertTrue((generated_dir / "foo" / "Main.py").exists())

        python = run_python(["-c", "import foo.Main as m; print(m.answer())"], cwd=generated_dir)
        self.assertEqual(python.returncode, 0, python.stderr)
        self.assertEqual(python.stdout.strip(), "7")

    def test_compile_generate_python_accepts_object_oriented_modules(self):
        root = self.temp_project()
        source_dir = root / "src"
        generated_dir = root / "generated"
        (source_dir / "foo").mkdir(parents=True)
        (source_dir / "foo" / "Main.coo").write_text(textwrap.dedent("""
            class Main {
                def main(args: List[String]): int = args.size()

                def inverted(flag: bool): bool = !flag

                def not_one(value: int): bool = value != 1
            }
        """))

        result = run_capy(["compile-generate", "py", "-i", str(source_dir), "-o", str(generated_dir)])

        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertTrue((generated_dir / "foo" / "Main.py").exists())

        python = run_python([str(generated_dir / "foo" / "Main.py"), "one", "two"])
        self.assertEqual(python.returncode, 0, python.stderr)
        self.assertEqual(python.stdout.strip(), "2")

        python = run_python([
            "-c",
            "from foo.Main import Main; m = Main(); "
            "print('|'.join(["
            "str(m.inverted(False)).lower(), "
            "str(m.not_one(2)).lower(), "
            "str(m.not_one(1)).lower()"
            "]))",
        ], cwd=generated_dir)
        self.assertEqual(python.returncode, 0, python.stderr)
        self.assertEqual(python.stdout.strip(), "true|true|false")

    def test_python_capybara_runner_applies_test_selectors(self):
        root = self.temp_project()
        generated_dir = root / "generated"
        output_dir = root / "results"
        runtime_dir = generated_dir / "capy" / "test"
        executed_file = generated_dir / "executed.txt"
        runtime_dir.mkdir(parents=True)
        (runtime_dir / "CapyTestRuntime.py").write_text(textwrap.dedent(f"""
            import pathlib

            executed_file = pathlib.Path({str(executed_file)!r})

            def record(name):
                def body():
                    with executed_file.open("a") as output:
                        output.write(name + "\\n")
                    return {{"result": True, "message": "", "type": "Assertion"}}
                return body

            def gatherTests():
                return [
                    {{
                        "file_name": "/capy/lang/EffectTest.cfun",
                        "test_cases": [
                            {{"name": "selected", "body": record("selected")}},
                            {{"name": "skipped", "body": record("skipped")}},
                        ],
                    }},
                    {{
                        "file_name": "/capy/lang/StringTest.cfun",
                        "test_cases": [
                            {{"name": "other", "body": record("other")}},
                        ],
                    }},
                ]
        """))
        runner = (
            pathlib.Path(__file__).resolve().parents[4]
            / "lib"
            / "capybara-lib"
            / "src"
            / "test"
            / "py"
            / "run-capybara-tests.py"
        )

        result = run_python([
            str(runner),
            "--generated-dir", str(generated_dir),
            "--output-dir", str(output_dir),
            "--tests", "/capy/lang/EffectTest.\"selected\"",
        ])

        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(executed_file.read_text().splitlines(), ["selected"])
        report = (output_dir / "TEST-capy.lang.EffectTest.cfun.xml").read_text()
        self.assertIn('tests="1"', report)
        self.assertIn('name="selected"', report)
        self.assertNotIn('name="skipped"', report)

        missing = run_python([
            str(runner),
            "--generated-dir", str(generated_dir),
            "--output-dir", str(root / "missing-results"),
            "--tests", "/capy/lang/EffectTest.\"missing\"",
        ])

        self.assertNotEqual(missing.returncode, 0)
        self.assertIn(
            "Test selector `/capy/lang/EffectTest.\"missing\"` did not match any test",
            missing.stderr,
        )

    def test_generated_python_output_is_pruned_by_manifest_on_reuse(self):
        root = self.temp_project()
        source_dir = root / "src"
        generated_dir = root / "generated"
        (source_dir / "foo").mkdir(parents=True)
        (generated_dir / "stale").mkdir(parents=True)
        (source_dir / "foo" / "Main.cfun").write_text("fun answer(): int = 1\n")
        (generated_dir / "stale" / "Old.py").write_text("x = 1\n")

        result = run_capy(["compile-generate", "PYTHON", "-i", str(source_dir), "-o", str(generated_dir)])

        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertTrue((generated_dir / "foo" / "Main.py").exists())
        self.assertFalse((generated_dir / "stale" / "Old.py").exists())
        manifest = (generated_dir / ".capy-output-manifest").read_text()
        self.assertIn("foo/Main.py", manifest)

    def tearDown(self):
        # Tests use mkdtemp because subprocesses may still have files open briefly on Windows.
        pass
